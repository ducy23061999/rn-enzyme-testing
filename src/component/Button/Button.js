import React, { Component } from 'react';
import {
    View, 
    TouchableOpacity,
    Text,
    Linking
} from 'react-native'
import styles from './ButtonStyle'

export default function Button({onPress, text, url}) {
    const activeColor = onPress ? '#3432a8' : '#989d9e';

    const action = {
        onPressHandle: () => {
            if (url) {
                Linking.openURL(url);
                return;
            }
            onPress();
        }
    }

    return (
        <TouchableOpacity 
            style = {[
                styles.container,
                {backgroundColor: activeColor}
            ]}
            onPress = {action.onPressHandle}
        >
            <Text style = {styles.text}>{text}</Text>
        </TouchableOpacity>
    )
}
