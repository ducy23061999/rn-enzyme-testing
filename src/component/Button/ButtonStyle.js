import {StyleSheet} from 'react-native'
import layouts from '../../layout'

export default StyleSheet.create({
    container: {
        height: 60,
        borderRadius: 10,
        ...layouts.rowCenter
    },
    text: {
        color: 'white',
        fontSize: 20
    }
})