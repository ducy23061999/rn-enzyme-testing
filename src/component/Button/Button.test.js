import React from 'react'
import {shallow, mount, render} from 'enzyme'
import Button from './Button'
// import {Linking} from 'react-native'

const mockOpenURL = jest.fn();
    
jest.mock('react-native/Libraries/Linking/Linking', () => ({
    openURL: mockOpenURL
}))

describe('Button', () => {
    describe('Rendering', () => {
        it('should be match in snapshot', () => {
            const component = shallow(
                <Button 
                    text ='Click me' 
                    onPress = {() => {}}
                />
            )
            expect(component).toMatchSnapshot()
        })
    })
    
    describe('Expect handle call once', () => {
        it('call without url', () => {
            const mockOnPress = jest.fn();

            const component = shallow(
                <Button 
                    text ='Click me' 
                    onPress = {mockOnPress}
                />
            )
            const instance = component.instance();

            // instance.drive().simulate("press")
            component.simulate("press")
            expect(mockOpenURL).not.toHaveBeenCalled();
            expect(mockOpenURL).toHaveBeenCalledTimes(0);
            expect(mockOnPress).toHaveBeenCalled();
            expect(mockOnPress).toHaveBeenCalledTimes(1);
        })

        it('call with url', () => {
            const mockOnPress = jest.fn();

            const component = shallow(
                <Button 
                    text ='Click me' 
                    onPress = {mockOnPress}
                    url = "https://www.google.com/"
                />
            )

            component.simulate("press")

            expect(mockOpenURL).toHaveBeenCalled();
            expect(mockOpenURL).toHaveBeenCalledTimes(1);
            expect(mockOnPress).not.toHaveBeenCalled();
            expect(mockOnPress).toHaveBeenCalledTimes(0);
        })
    })
})