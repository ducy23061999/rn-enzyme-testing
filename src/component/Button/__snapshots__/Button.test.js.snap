// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Rendering should be match in snapshot 1`] = `
<ForwardRef
  onPress={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 10,
        "display": "flex",
        "flexDirection": "row",
        "height": 60,
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#3432a8",
      },
    ]
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 20,
      }
    }
  >
    Click me
  </Text>
</ForwardRef>
`;
